packages/server/modules/serverinvites/services/creation.ts

```diff
@@ -0,0 +1,215 @@
+) => {
+  if (targetUser) target = buildUserTarget(targetUser.id)!
+  return resolveTarget(target)
+}
+
+/**
+ * Sanitize message that potentially has HTML in it
+ */
+function sanitizeMessage(message: string, stripAll: boolean = false) {
+  return sanitizeHtml(message, {
+    allowedTags: stripAll ? [] : ['b', 'i', 'em', 'strong']
+  })
+}
+
+const sendInviteEmailFactory =
+  (deps: { buildInviteEmailContents: BuildInviteEmailContents }) =>
+  async (params: {
+    invite: ServerInviteRecord
+    inviter: UserWithOptionalRole
+    serverInfo: ServerInfo
+    targetUser?: MaybeNullOrUndefined<UserWithOptionalRole>
+    targetData: ResolvedTargetData
+  }) => {
+    const { invite, inviter, serverInfo, targetUser, targetData } = params
+
+    const emailContents = await deps.buildInviteEmailContents({
+      invite,
+      inviter,
+      serverInfo
+    })
+    const renderedEmail = await renderEmail(
+      emailContents.emailParams,
+      serverInfo,
+      targetUser || null
+    )
+
+    // send email and emit event
+    await emailsModule.sendEmail({
+      subject: emailContents.subject,
+      to: targetUser ? targetUser.email : targetData.userEmail!,
+      ...renderedEmail
+    })
+  }
+
+/**
+ * Create and send out an invite
+ */
+export const createAndSendInviteFactory =
+  ({
+    findUserByTarget,
+    insertInviteAndDeleteOld,
+    collectAndValidateResourceTargets,
+    buildInviteEmailContents,
+    emitServerInvitesEvent
+  }: {
+    findUserByTarget: FindUserByTarget
+    insertInviteAndDeleteOld: InsertInviteAndDeleteOld
+    collectAndValidateResourceTargets: CollectAndValidateResourceTargets
+    buildInviteEmailContents: BuildInviteEmailContents
+    emitServerInvitesEvent: EmitServerInvitesEvent
+  }): CreateAndSendInvite =>
+  async (params, inviterResourceAccessLimits?) => {
+    const sendInviteEmail = sendInviteEmailFactory({ buildInviteEmailContents })
+    const { inviterId, target } = params
+    let { message } = params
+
+    const [inviter, targetUser, serverInfo] = await Promise.all([
+      getUser(inviterId, { withRole: true }),
+      findUserByTarget(target),
+      getServerInfo()
+    ])
+    if (!inviter) throw new InviteCreateValidationError('Invalid inviter')
+
+    // if target user found, always use the user ID
+    const targetData = getFinalTargetData(target, targetUser)
+    if (targetData.userId && !targetUser) {
+      throw new InviteCreateValidationError('Attempting to invite an invalid user')
+    }
+
+    if (message && message.length >= 1024) {
+      throw new InviteCreateValidationError('Personal message too long')
+    }
+
+    // collect and validate all resource targets
+    const resources = await collectAndValidateResourceTargets({
+      input: params,
+      inviter,
+      inviterResourceAccessLimits,
+      target: targetData,
+      targetUser,
+      serverInfo
+    })
+    const finalPrimaryResource = getPrimaryResourceTarget(resources)
```

Comment:
cc Gergo: It doesn't make any sense to save all collected resource targets in the DB. Not only does that complicate SQL queries, but it also enables a situation where we decide that project invites also should invite people to a new type of resource, but all of pre-existing project invites won't do it, because they resolved their targets at the moment of invite creation, not when accepting it